lightningd/peer_htlcs.c

```diff
@@ -677,29 +751,53 @@ htlc_accepted_hook_callback(struct htlc_accepted_hook_payload *request,
 	struct channel *channel = request->channel;
 	struct lightningd *ld = request->ld;
 	u8 *req;
+	enum htlc_accepted_result result;
+	struct htlc_accepted_hook_response *response;
+	response = htlc_accepted_hook_deserialize(request, buffer, toks);
 
-	/* TODO(cdecker) Assign to *response once we actually use it */
-	htlc_accepted_hook_deserialize(request, buffer, toks);
-
-	if (rs->nextcase == ONION_FORWARD) {
-		struct gossip_resolve *gr = tal(ld, struct gossip_resolve);
-
-		gr->next_onion = tal_steal(gr, request->next_onion); serialize_onionpacket(gr, rs->next);
-		gr->next_channel = rs->hop_data.channel_id;
-		gr->amt_to_forward = rs->hop_data.amt_forward;
-		gr->outgoing_cltv_value = rs->hop_data.outgoing_cltv;
-		gr->hin = hin;
+	if (response)
+		result = response->result;
+	else
+		result = htlc_accepted_continue;
+
+	switch (result) {
+	case htlc_accepted_continue:
+		if (rs->nextcase == ONION_FORWARD) {
+			struct gossip_resolve *gr =
+			    tal(ld, struct gossip_resolve);
+
+			gr->next_onion = tal_steal(gr, request->next_onion);
+			serialize_onionpacket(gr, rs->next);
+			gr->next_channel = rs->hop_data.channel_id;
+			gr->amt_to_forward = rs->hop_data.amt_forward;
+			gr->outgoing_cltv_value = rs->hop_data.outgoing_cltv;
+			gr->hin = hin;
+
+			req = towire_gossip_get_channel_peer(tmpctx,
+							     &gr->next_channel);
+			log_debug(
+			    channel->log, "Asking gossip to resolve channel %s",
+			    type_to_string(tmpctx, struct short_channel_id,
+					   &gr->next_channel));
+			subd_req(hin, ld->gossip, req, -1, 0,
+				 channel_resolve_reply, gr);
+		} else
+			handle_localpay(hin, hin->cltv_expiry,
+					&hin->payment_hash,
+					rs->hop_data.amt_forward,
+					rs->hop_data.outgoing_cltv);
+
+		break;
+	case htlc_accepted_fail:
+		log_debug(channel->log,
+			  "Failing incoming HTLC as instructed by plugin hook");
+		fail_in_htlc(hin, response->failure_code, NULL, NULL);
+		break;
+	case htlc_accepted_resolve:
+		fulfill_htlc(hin, &response->payment_key);
+		break;
```

Comment:
You realize this means we can write a  "try_stealing" plugin now that simply fulfills known preimages?  :)

Reply:
And this is a problem? :)

Reply:
Well, that's my evil master plan thwarted :stuck_out_tongue_winking_eye: